managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java

```diff
@@ -323,6 +350,72 @@ void asyncReadEntry0(ReadHandle lh, long firstEntry, long lastEntry, boolean sho
         }
     }
 
+    private AsyncCallbacks.ReadEntriesCallback handlePendingReadsLimits(ReadHandle lh,
+                                                                long firstEntry, long lastEntry,
+                                                                boolean shouldCacheEntry,
+                                                                AsyncCallbacks.ReadEntriesCallback originalCallback,
+                                                                Object ctx, PendingReadsLimiter.Handle handle) {
+        PendingReadsLimiter pendingReadsLimiter = getPendingReadsLimiter();
+        if (pendingReadsLimiter.isDisabled()) {
+            return originalCallback;
+        }
+        long estimatedReadSize = (1 + lastEntry - firstEntry)
+                * (estimatedEntrySize + BOOKKEEPER_READ_OVERHEAD_PER_ENTRY);
+        final AsyncCallbacks.ReadEntriesCallback callback;
+        PendingReadsLimiter.Handle newHandle = pendingReadsLimiter.acquire(estimatedReadSize, handle);
+        if (!newHandle.success) {
+            long now = System.currentTimeMillis();
+            if (now - newHandle.creationTime > readEntryTimeoutMillis) {
+                String message = "Time-out elapsed reading from ledger "
+                        + lh.getId()
+                        + ", " + getName()
+                        + ", estimated read size " + estimatedReadSize + " bytes"
+                        + " for " + (1 + lastEntry - firstEntry) + " entries";
+                log.error(message);
+                pendingReadsLimiter.release(newHandle);
+                originalCallback.readEntriesFailed(
+                        new ManagedLedgerException.TooManyRequestsException(message), ctx);
+                return null;
+            }
+            ml.getExecutor().submitOrdered(lh.getId(), () -> {
+                asyncReadEntry0WithLimits(lh, firstEntry, lastEntry, shouldCacheEntry,
+                        originalCallback, ctx, newHandle);
+                return null;
+            });
+            return null;
+        } else {
+            callback = new AsyncCallbacks.ReadEntriesCallback() {
+
+                @Override
+                public void readEntriesComplete(List<Entry> entries, Object ctx) {
+                    if (!entries.isEmpty()) {
```

Comment:
We now know the exact size of the entries, if we defer releasing the permits, should we also readjust the number of permits granted? That would give us a more accurate limiter here.

Reply:
I have thought about this and finally I decided to keep it simple.
You have two cases:
1) you acquired more permits than necessary -> we could early release some permits -> this is SIMPLE
2) you acquired less permits than necessary -> PROBLEM ! we cannot acquire more permits here

We cannot do 2), it is too late here
Doing only 1) without 2) seems overkilling and not consistent. The benefit would be to give more chances to other competing reads to move forward earlier. I don't think it is worth

@michaeljmarshall WDYT?

Reply:
I agree that 1 is simple, and my response for number 2 is that we could allow for negative permits. The main benefit is that on the request side, we'll back pressure for a little longer until the permits go into the positive territory.

I also agree with you that doing 1 without 2 is not a good solution since our error will always be to underestimate the size of the in memory data.

I am fine with leaving this as a potential optimization, if you would prefer to keep it simple here. As long as we are getting a good enough average of entries, we should be fine.